# KNN Classifier atau Klasifikasi

**Nama  : Rizki Ardian Samudra**

**NIM   : 210411100179**

**Kelas : Penambangan Data-C**

KNN pada dasarnya adalah algoritma supervised machine learning. Dalam bahasa yang lebih sederhana ialah mencari pola dalam data yang ada untuk membuat prediksi. Prediksi tersebut bisa menjadi memprediksi harga mobil atau mengidentifikasi pelanggan mana yang akan merespons dengan baik pada kampanye pemasaran email baru.

Jadi KNN dalam bentuk yang paling sederhana adalah mengidentifikasi titik data yang memiliki fitur yang mirip dengan titik data Anda, menuliskannya berdasarkan kemiripan (similarity), dan memilih k jumlah akhir dari daftar tersebut. Prediksi kemudian dapat dilakukan dengan mengambil rata-rata atau mean dari fitur dari titik akhir. Ini pada dasarnya adalah alur kerja dari algoritma KNN.

Jadi nanti Kita menghitung kemiripan (similarity) menggunakan apa yang kita sebut sebagai metrik kemiripan. Metrik kemiripan membandingkan satu set fitur numerik tetap antara dua pengamatan.

1. Mengimport library yang diperlukan

In [ ]:
import pandas as pd
import numpy as np
from sklearn.neighbors import KNeighborsRegressor 
from sklearn.preprocessing import LabelEncoder

from sklearn.impute import KNNImputer
from sklearn.preprocessing import OrdinalEncoder


2. Import data kita

In [ ]:
df = pd.read_csv('https://raw.githubusercontent.com/datasets/breast-cancer/master/data/breast-cancer.csv')

jumlah_baris = df.shape[0]
df.head(jumlah_baris)

age mefalsepause tumor-size inv-falsedes falsede-caps  deg-malig  \
0    40-49   premefalse      15-19          0-2         True          3   
1    50-59         ge40      15-19          0-2        False          1   
2    50-59         ge40      35-39          0-2        False          2   
3    40-49   premefalse      35-39          0-2         True          3   
4    40-49   premefalse      30-34          3-5         True          2   
..     ...          ...        ...          ...          ...        ...   
267  50-59         ge40      30-34          6-8         True          2   
268  50-59   premefalse      25-29          3-5         True          2   
269  30-39   premefalse      30-34          6-8         True          2   
270  50-59   premefalse      15-19          0-2        False          2   
271  50-59         ge40      40-44          0-2        False          3   

    breast breast-quad  irradiat                    class  
0    right     left_up     False        recurrence-events  
1    right     central     False  false-recurrence-events  
2     left    left_low     False        recurrence-events  
3    right    left_low      True  false-recurrence-events  
4     left    right_up     False        recurrence-events  
..     ...         ...       ...                      ...  
267   left    left_low     False  false-recurrence-events  
268   left    left_low      True  false-recurrence-events  
269  right    right_up     False  false-recurrence-events  
270  right    left_low     False  false-recurrence-events  
271   left    right_up     False  false-recurrence-events  

[272 rows x 10 columns]

In [ ]:
df = df.rename(columns={'tumor-size': 'tumor_size'})
df = df.rename(columns={'inv-falsedes': 'inv_falsedes'})
df = df.rename(columns={'falsede-caps': 'falsede_caps'})
df = df.rename(columns={'deg-malig': 'deg_malig'})
df = df.rename(columns={'breast-quad': 'breast_quad'})
df = df.rename(columns={'inv-falsedes': 'inv_falsedes'})
df = df.rename(columns={'class': 'class_name'})
df.head(jumlah_baris)

age mefalsepause tumor_size inv_falsedes falsede_caps  deg_malig  \
0    40-49   premefalse      15-19          0-2         True          3   
1    50-59         ge40      15-19          0-2        False          1   
2    50-59         ge40      35-39          0-2        False          2   
3    40-49   premefalse      35-39          0-2         True          3   
4    40-49   premefalse      30-34          3-5         True          2   
..     ...          ...        ...          ...          ...        ...   
267  50-59         ge40      30-34          6-8         True          2   
268  50-59   premefalse      25-29          3-5         True          2   
269  30-39   premefalse      30-34          6-8         True          2   
270  50-59   premefalse      15-19          0-2        False          2   
271  50-59         ge40      40-44          0-2        False          3   

    breast breast_quad  irradiat               class_name  
0    right     left_up     False        recurrence-events  
1    right     central     False  false-recurrence-events  
2     left    left_low     False        recurrence-events  
3    right    left_low      True  false-recurrence-events  
4     left    right_up     False        recurrence-events  
..     ...         ...       ...                      ...  
267   left    left_low     False  false-recurrence-events  
268   left    left_low      True  false-recurrence-events  
269  right    right_up     False  false-recurrence-events  
270  right    left_low     False  false-recurrence-events  
271   left    right_up     False  false-recurrence-events  

[272 rows x 10 columns]

3. Mengecek nilai yang hilang/kosong pada data

In [ ]:
df_copy = df.copy()
df_copy.isnull().sum()

age             0
mefalsepause    0
tumor_size      0
inv_falsedes    0
falsede_caps    8
deg_malig       0
breast          0
breast_quad     1
irradiat        0
class_name      0
dtype: int64

4. Ganti nilai yang mengandung tanda tanya (?) menjadi nilai kosong

In [ ]:
df_copy = df_copy.replace('?',np.nan)

5. Cek lagi berapa nilai yang kosong pada data

In [ ]:
df_copy.isnull().sum()

age             0
mefalsepause    0
tumor_size      0
inv_falsedes    0
falsede_caps    8
deg_malig       0
breast          0
breast_quad     1
irradiat        0
class_name      0
dtype: int64

6. Setelah itu kita lihat type datanya, jika type datanya bukan numerik/float maka perlu diganti type datanya

In [ ]:
df_copy.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 272 entries, 0 to 271
Data columns (total 10 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   age           272 non-null    object
 1   mefalsepause  272 non-null    object
 2   tumor_size    272 non-null    object
 3   inv_falsedes  272 non-null    object
 4   falsede_caps  264 non-null    object
 5   deg_malig     272 non-null    int64 
 6   breast        272 non-null    object
 7   breast_quad   271 non-null    object
 8   irradiat      272 non-null    bool  
 9   class_name    272 non-null    object
dtypes: bool(1), int64(1), object(8)
memory usage: 19.5+ KB


7. Mengubah semuanya tipe data pada fitur

In [ ]:
df_copy['deg_malig'] = df_copy['deg_malig'].astype('object')
df_copy.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 272 entries, 0 to 271
Data columns (total 10 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   age           272 non-null    object
 1   mefalsepause  272 non-null    object
 2   tumor_size    272 non-null    object
 3   inv_falsedes  272 non-null    object
 4   falsede_caps  264 non-null    object
 5   deg_malig     272 non-null    object
 6   breast        272 non-null    object
 7   breast_quad   271 non-null    object
 8   irradiat      272 non-null    bool  
 9   class_name    272 non-null    object
dtypes: bool(1), object(9)
memory usage: 19.5+ KB


In [ ]:
df_copy.describe(include= 'all')

age mefalsepause tumor_size inv_falsedes falsede_caps  deg_malig  \
count     272          272        272          272          264        272   
unique      6            3         11            7            2          3   
top     50-59   premefalse      30-34          0-2        False          2   
freq       91          143         60          200          209        124   

       breast breast_quad irradiat               class_name  
count     272         271      272                      272  
unique      2           5        2                        2  
top      left    left_low    False  false-recurrence-events  
freq      143         103      205                      191

In [ ]:
for value in ['falsede_caps', 'breast_quad']:
    df_copy[value].replace([None], [df_copy.describe(include='all')[value][2]], inplace=True)  #mengisi berdasarkan top describe

df_copy.head(jumlah_baris)

age mefalsepause tumor_size inv_falsedes  falsede_caps deg_malig  \
0    40-49   premefalse      15-19          0-2          True         3   
1    50-59         ge40      15-19          0-2         False         1   
2    50-59         ge40      35-39          0-2         False         2   
3    40-49   premefalse      35-39          0-2          True         3   
4    40-49   premefalse      30-34          3-5          True         2   
..     ...          ...        ...          ...           ...       ...   
267  50-59         ge40      30-34          6-8          True         2   
268  50-59   premefalse      25-29          3-5          True         2   
269  30-39   premefalse      30-34          6-8          True         2   
270  50-59   premefalse      15-19          0-2         False         2   
271  50-59         ge40      40-44          0-2         False         3   

    breast breast_quad  irradiat               class_name  
0    right     left_up     False        recurrence-events  
1    right     central     False  false-recurrence-events  
2     left    left_low     False        recurrence-events  
3    right    left_low      True  false-recurrence-events  
4     left    right_up     False        recurrence-events  
..     ...         ...       ...                      ...  
267   left    left_low     False  false-recurrence-events  
268   left    left_low      True  false-recurrence-events  
269  right    right_up     False  false-recurrence-events  
270  right    left_low     False  false-recurrence-events  
271   left    right_up     False  false-recurrence-events  

[272 rows x 10 columns]

10. Cek apakah masih ada data yang kosong

In [ ]:
df_copy.isnull().sum()

age             0
mefalsepause    0
tumor_size      0
inv_falsedes    0
falsede_caps    0
deg_malig       0
breast          0
breast_quad     0
irradiat        0
class_name      0
dtype: int64

11. Sekarang coba kita tampilkan data yang terbaru

In [ ]:
df_copy.head(jumlah_baris)

age mefalsepause tumor_size inv_falsedes  falsede_caps deg_malig  \
0    40-49   premefalse      15-19          0-2          True         3   
1    50-59         ge40      15-19          0-2         False         1   
2    50-59         ge40      35-39          0-2         False         2   
3    40-49   premefalse      35-39          0-2          True         3   
4    40-49   premefalse      30-34          3-5          True         2   
..     ...          ...        ...          ...           ...       ...   
267  50-59         ge40      30-34          6-8          True         2   
268  50-59   premefalse      25-29          3-5          True         2   
269  30-39   premefalse      30-34          6-8          True         2   
270  50-59   premefalse      15-19          0-2         False         2   
271  50-59         ge40      40-44          0-2         False         3   

    breast breast_quad  irradiat               class_name  
0    right     left_up     False        recurrence-events  
1    right     central     False  false-recurrence-events  
2     left    left_low     False        recurrence-events  
3    right    left_low      True  false-recurrence-events  
4     left    right_up     False        recurrence-events  
..     ...         ...       ...                      ...  
267   left    left_low     False  false-recurrence-events  
268   left    left_low      True  false-recurrence-events  
269  right    right_up     False  false-recurrence-events  
270  right    left_low     False  false-recurrence-events  
271   left    right_up     False  false-recurrence-events  

[272 rows x 10 columns]

In [ ]:
%%capture

#instantiate both packages to use
encoder = OrdinalEncoder()
imputer = KNNImputer()
# create a list of categorical columns to iterate over
cat_cols = ['age', 'mefalsepause', 'tumor_size', 'inv_falsedes', 'falsede_caps', 'deg_malig', 'breast', 'breast_quad', 'irradiat', 'class_name']

def encode(data):
    '''function to encode non-null data and replace it in the original data'''
    #retains only non-null values
    nonulls = np.array(data.dropna())
    #reshapes the data for encoding
    impute_reshape = nonulls.reshape(-1,1)
    #encode date
    impute_ordinal = encoder.fit_transform(impute_reshape)
    #Assign back encoded values to non-null values
    data.loc[data.notnull()] = np.squeeze(impute_ordinal)
    return data

#create a for loop to iterate through each column in the data
for columns in cat_cols:
    encode(df_copy[columns])

In [ ]:
df_copy.head(jumlah_baris)

age mefalsepause tumor_size inv_falsedes falsede_caps deg_malig breast  \
0    2.0          2.0        2.0          0.0          1.0       2.0    1.0   
1    3.0          0.0        2.0          0.0          0.0       0.0    1.0   
2    3.0          0.0        6.0          0.0          0.0       1.0    0.0   
3    2.0          2.0        6.0          0.0          1.0       2.0    1.0   
4    2.0          2.0        5.0          4.0          1.0       1.0    0.0   
..   ...          ...        ...          ...          ...       ...    ...   
267  3.0          0.0        5.0          5.0          1.0       1.0    0.0   
268  3.0          2.0        4.0          4.0          1.0       1.0    0.0   
269  1.0          2.0        5.0          5.0          1.0       1.0    1.0   
270  3.0          2.0        2.0          0.0          0.0       1.0    1.0   
271  3.0          0.0        7.0          0.0          0.0       2.0    0.0   

    breast_quad irradiat class_name  
0           2.0      0.0        1.0  
1           0.0      0.0        0.0  
2           1.0      0.0        1.0  
3           1.0      1.0        0.0  
4           4.0      0.0        1.0  
..          ...      ...        ...  
267         1.0      0.0        0.0  
268         1.0      1.0        0.0  
269         4.0      0.0        0.0  
270         1.0      0.0        0.0  
271         4.0      0.0        0.0  

[272 rows x 10 columns]

12. Melakukan normalisasi data, yaitu mengubah setiap nilai ke dalam rentang antara 0 hingga 1. Kita akan menggunakan metode min-max, yaitu: x - min / max - min. Terdapat juga rumus-rumus lain yang dapat digunakan, misalnya metode simple feature scaling atau metode z-score. Namun perlu diingat bahwa kolom target tidak akan dinormalisasi (kolom species).

In [ ]:
class_name_column = df_copy['class_name']
df_copy = df_copy[['age', 'mefalsepause', 'tumor_size', 'inv_falsedes', 'falsede_caps', 'deg_malig', 'breast', 'breast_quad', 'irradiat']].copy()
df_copy = (df_copy - df_copy.min())/(df_copy.max() - df_copy.min())
df_copy['class_name'] = class_name_column.copy()
df_copy.head(jumlah_baris)

age mefalsepause tumor_size inv_falsedes falsede_caps deg_malig breast  \
0    0.4          1.0        0.2          0.0          1.0       1.0    1.0   
1    0.6          0.0        0.2          0.0          0.0       0.0    1.0   
2    0.6          0.0        0.6          0.0          0.0       0.5    0.0   
3    0.4          1.0        0.6          0.0          1.0       1.0    1.0   
4    0.4          1.0        0.5     0.666667          1.0       0.5    0.0   
..   ...          ...        ...          ...          ...       ...    ...   
267  0.6          0.0        0.5     0.833333          1.0       0.5    0.0   
268  0.6          1.0        0.4     0.666667          1.0       0.5    0.0   
269  0.2          1.0        0.5     0.833333          1.0       0.5    1.0   
270  0.6          1.0        0.2          0.0          0.0       0.5    1.0   
271  0.6          0.0        0.7          0.0          0.0       1.0    0.0   

    breast_quad irradiat class_name  
0           0.5      0.0        1.0  
1           0.0      0.0        0.0  
2          0.25      0.0        1.0  
3          0.25      1.0        0.0  
4           1.0      0.0        1.0  
..          ...      ...        ...  
267        0.25      0.0        0.0  
268        0.25      1.0        0.0  
269         1.0      0.0        0.0  
270        0.25      0.0        0.0  
271         1.0      0.0        0.0  

[272 rows x 10 columns]

13. Terakhir Kita menginisialisasi objek knn dari kelas KNeighborsRegressor yang merupakan algoritma K-Nearest Neighbors untuk melakukan regresi. Kemudian, kita melakukan randomisasi indeks dari dataset menggunakan np.random.permutation() dengan seed 1 untuk memastikan hasil yang konsisten.

> Selanjutnya, kita membagi dataset menjadi 75% data untuk training dan 25% untuk testing menggunakan variabel train_index. Data training disimpan di training_data, dan data testing disimpan di testing_data.

> Setelah itu, kita melakukan training model menggunakan data training dengan memanggil method fit() pada objek knn dengan argumen input berupa fitur-fitur yang digunakan (dalam hal ini sepal_length, sepal_width, petal_length, dan petal_width) dan label yaitu species.

> Kemudian, kita melakukan prediksi pada data testing dengan memanggil method predict() pada objek knn dengan argumen input berupa fitur-fitur dari data testing yang ingin diprediksi.

> Setelah itu, kita menghitung akurasi model dengan mengukur berapa persen data testing yang berhasil diprediksi dengan (threshold default 0.5.).

> Terakhir, kita mencetak hasil akurasi model.

In [ ]:
knn = KNeighborsRegressor()

# before we split, make sure you randomize the dataset
np.random.seed(1)
index_shuffled = np.random.permutation(df_copy.index)
shuffled_car_data = df_copy.loc[index_shuffled]

# We will use 75% of the data for training and the remaining 25% for testing
train_index = int(shuffled_car_data.shape[0] * 0.75)
training_data = shuffled_car_data.iloc[0:train_index]
testing_data = shuffled_car_data.iloc[train_index:]

# fit the model with the training data
knn.fit(training_data[['age', 'mefalsepause', 'tumor_size', 'inv_falsedes', 'falsede_caps', 'deg_malig', 'breast', 'breast_quad', 'irradiat']], training_data['class_name'])

# predict the species of the testing data
prediction = knn.predict(testing_data[['age', 'mefalsepause', 'tumor_size', 'inv_falsedes', 'falsede_caps', 'deg_malig', 'breast', 'breast_quad', 'irradiat']])

# calculate the accuracy of the model
accuracy = np.mean(np.abs(prediction - testing_data['class_name']) <= 0.5)

print(f"Akurasi model KNN: {accuracy:.2%}")


Akurasi model KNN: 77.94%


## Kesimpulan

Berdasarkan hasil diatas, tingkat akurasi yang diperoleh dari model KNN adalah 77.94%